### PR TITLE
Discover workspaces without using them in resolution

### PR DESCRIPTION
**Pull request by konstin**
Add minimal support for workspace discovery, only used for determining paths in the bluejay commands.

We can now discover the workspace structure, namely that the `pyproject.toml` of a package belongs to a workspace `pyproject.toml` with members and exclusion. The globbing logic is inspired by cargo. We don't resolve `workspace = true` metadata declarations yet.

### CODE REVIEW

File: crates/uv-requirements/src/discovery.rs [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+        let project = ProjectWorkspace::discover(root_dir.join(folder)).unwrap();
+        let root_escaped = regex::escape(root_dir.to_string_lossy().as_ref());
+        (project, root_escaped)
+    }
+
+    #[test]
+    fn albatross_in_example() {
+        let (project, root_escaped) = workspace_test("albatross-root-workspace");
+        let filters = vec![(root_escaped.as_str(), "[ROOT]")];
+        insta::with_settings!({filters => filters}, {
```
Comment (konstin) [inline crates/uv-requirements/src/discovery.rs:418]: I tried the json snapshots here for the first time and find them better than the regular debug snapshots

File: crates/uv-requirements/src/pyproject.rs [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+//! Reads the following fields from `pyproject.toml`:
 //!
 //! * `project.{dependencies,optional-dependencies}`
 //! * `tool.uv.sources`
 //! * `tool.uv.workspace`
 //!
 //! Then lowers them into a dependency specification.
 
-use std::collections::HashMap;
+use std::collections::BTreeMap;
```
Comment (konstin) [inline crates/uv-requirements/src/pyproject.rs:9]: This makes the test snapshots deterministic

File: crates/uv-requirements/src/discovery.rs
Comment (konstin) [inline crates/uv-requirements/src/discovery.rs:1]: See for a more used state of this file https://gist.github.com/konstin/f8c05311223fbd627b0491295915e865

File: crates/uv-requirements/src/discovery.rs [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+}
+
+/// A package in a workspace.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(serde::Serialize))]
+#[allow(dead_code)] // TODO(konsti): Resolve workspace package declarations.
+pub struct WorkspaceMember {
+    /// The path to the project root.
+    root: PathBuf,
+    pyproject_toml: PyProjectToml,
```
Comment (zanieb) [inline crates/uv-requirements/src/discovery.rs:42]: Does this need a doc too?

File: crates/uv-requirements/src/discovery.rs [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+    pub fn discover(path: impl AsRef<Path>) -> Result<Self, DiscoverError> {
+        let Some(project_root) = path
+            .as_ref()
+            .ancestors()
+            .find(|path| path.join("pyproject.toml").exists())
+        else {
+            return Err(DiscoverError::MissingPyprojectToml);
+        };
+
+        debug!("Project root: `{}`", project_root.simplified_display());
```
Comment (zanieb) [inline crates/uv-requirements/src/discovery.rs:73]: Can you make this message a bit more user-facing e.g. "Found project root `{}`" or similar?

File: crates/uv-requirements/src/discovery.rs [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+
+        Self::from_project_root(project_root)
+    }
+
+    /// The directory containing the closest `pyproject.toml`, containing a `[project]` section.
+    pub fn project_root(&self) -> &Path {
+        &self.project_root
+    }
+
+    pub fn project_name(&self) -> &PackageName {
```
Comment (zanieb) [inline crates/uv-requirements/src/discovery.rs:83]: Does this need a doc? Why "project name" vs "package name"?

File: crates/uv-requirements/src/discovery.rs [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+    workspace_sources: BTreeMap<PackageName, Source>,
+}
+
+impl ProjectWorkspace {
+    /// Find the current project and workspace.
+    pub fn discover(path: impl AsRef<Path>) -> Result<Self, DiscoverError> {
+        let Some(project_root) = path
+            .as_ref()
+            .ancestors()
+            .find(|path| path.join("pyproject.toml").exists())
```
Comment (zanieb) [inline crates/uv-requirements/src/discovery.rs:68]: Should we use `is_file()` instead of `exists()`?

**Review by zanieb — COMMENTED** [review]
Could you add some documentation explaining what's going on? Or at least link to the relevant "specification" we're implementing here? I'm having a hard time reviewing because there's a lot of behavior in here.

**Comment by konstin** [comment]
I've edited the spec from #3404 into a docstring

**Review by charliermarsh — COMMENTED** [review]
If this is causing you trouble, you can merge and I'll do a post-push review.

**Comment by konstin** [comment]
Merging this now to avoid merge conflicts

**Comment by charliermarsh** [comment]
@konstin - Gave this a read, looks like a good start.

File: crates/uv-requirements/src/workspace.rs [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+    ) -> Result<Self, WorkspaceError> {
+        let mut workspace = project
+            .tool
+            .as_ref()
+            .and_then(|tool| tool.uv.as_ref())
+            .and_then(|uv| uv.workspace.as_ref())
+            .map(|workspace| (project_path.clone(), workspace.clone(), project.clone()));
+
+        if workspace.is_none() {
+            workspace = find_workspace(&project_path)?;
```
Comment (charliermarsh) [inline crates/uv-requirements/src/workspace.rs:246]: Should this be an `or_else`?

Comment (charliermarsh) [reply crates/uv-requirements/src/workspace.rs:246]: Does this mean we're in a _member_ of a workspace?

Comment (konstin) [reply crates/uv-requirements/src/workspace.rs:246]: yeah, added some docs in the other branch

**Pull request merged**